Detectors/EMCAL/workflow/src/RawToCellConverterSpec.cxx

```diff
@@ -0,0 +1,149 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+#include "FairLogger.h"
+
+#include "DataFormatsEMCAL/EMCALBlockHeader.h"
+#include "DataFormatsEMCAL/TriggerRecord.h"
+#include "EMCALWorkflow/RawToCellConverterSpec.h"
+#include "Framework/ControlService.h"
+#include "SimulationDataFormat/MCCompLabel.h"
+#include "SimulationDataFormat/MCTruthContainer.h"
+#include "EMCALBase/Geometry.h"
+#include "EMCALBase/Mapper.h"
+#include "EMCALReconstruction/CaloFitResults.h"
+#include "EMCALReconstruction/Bunch.h"
+#include "EMCALReconstruction/CaloRawFitterStandard.h"
+#include "EMCALReconstruction/AltroDecoder.h"
+#include "CommonDataFormat/InteractionRecord.h"
+
+using namespace o2::emcal::reco_workflow;
+
+void RawToCellConverterSpec::init(framework::InitContext& ctx)
+{
+  LOG(DEBUG) << "[EMCALRawToCellConverter - init] Initialize converter ";
+  if (!mGeometry) {
+    mGeometry = o2::emcal::Geometry::GetInstanceFromRunNumber(223409);
+  }
+  if (!mGeometry)
+    LOG(ERROR) << "Failure accessing geometry";
+
+  if (!mMapper) {
+    mMapper = std::unique_ptr<o2::emcal::MappingHandler>(new o2::emcal::MappingHandler);
+  }
+  if (!mMapper)
+    LOG(ERROR) << "Failed to initialize mapper";
+
+  mRawFitter.setAmpCut(mNoiseThreshold);
+  mRawFitter.setL1Phase(0.);
+}
+
+void RawToCellConverterSpec::run(framework::ProcessingContext& ctx)
+{
+  LOG(DEBUG) << "[EMCALRawToCellConverter - run] called";
+  auto dataref = ctx.inputs().get("raw");
+  auto const* emcheader = o2::framework::DataRefUtils::getHeader<o2::emcal::EMCALBlockHeader*>(dataref);
+  if (!emcheader->mHasPayload) {
+    LOG(DEBUG) << "[EMCALCellConverter - run] No more raw data" << std::endl;
+    ctx.services().get<o2::framework::ControlService>().readyToQuit(framework::QuitRequest::Me);
+    return;
+  }
+
+  mOutputCells.clear();
+  mOutputTriggerRecords.clear();
+
+  int firstEntry = 0;
+  for (const auto& rawData : ctx.inputs()) {
+
+    //o2::emcal::RawReaderMemory<o2::header::RAWDataHeaderV4> rawreader(gsl::span(rawData.payload, o2::framework::DataRefUtils::getPayloadSize(rawData)));
+
+    o2::emcal::RawReaderMemory<o2::header::RAWDataHeaderV4> rawreader(o2::framework::DataRefUtils::as<const char>(rawData));
+
+    bool first = true;
+    uint16_t currentTrigger = 0;
+    uint32_t currentorbit = 0;
+
+    // loop over all the DMA pages
+    while (rawreader.hasNext()) {
+
+      rawreader.next();
+
+      o2::header::RAWDataHeaderV4 header = rawreader.getRawHeader();
```

Comment:
Please use `auto` in order to be compatible with future changes of the raw data header.